### PR TITLE
Add pre-commit hook to check session default value

### PR DESCRIPTION
**Pull request by uranusjr**
Is it surprising there are actually quite a few problematic usages in the code base?

### CODE REVIEW

File: .pre-commit-config.yaml [hunk 1]
```diff
@@ -424,6 +424,12 @@ repos:
         pass_filenames: true
         files: \.py$
         exclude: ^tests/|^airflow/_vendor/|^docs/
+      - id: check-only-new-session-with-provide-session
+        name: Check NEW_SESSION is only used with with @provide_session
```
Comment (pierrejeambrun) [inline .pre-commit-config.yaml]:
```suggestion
        name: Check NEW_SESSION is only used with @provide_session
```

File: scripts/ci/pre_commit/pre_commit_new_session_in_provide_session.py [hunk 2]
```diff
@@ -0,0 +1,123 @@
...
+        if default_kind == _SessionDefault.none and _annotation_has_none(session.argument.annotation):
+            continue  # None is OK if the argument is explicitly typed as None.
+        yield node
+
+
+def main(argv: list[str]) -> int:
+    paths = (pathlib.Path(filename) for filename in argv[1:])
+    errors = [(path, error) for path in paths for error in _iter_incorrect_new_session_usages(path)]
+    if errors:
+        print("Incorrect NEW_SESSION usages:", end="\n\n")
```
Comment (potiuk) [inline scripts/ci/pre_commit/pre_commit_new_session_in_provide_session.py]: I think we should explain the usages in detail - basically the comment you provided above should made it into message explaining what is wrong. The goal should be that those who made mistake and their static check fails should see what they should do to fix it. (rich-colorful if possible - I usually use Yellow to display this message in UI. Similarly the "Incorrect NEW_SESSION" shoud be in"[red]" - this way it is immediately visible in the CI output what is an error and what is the action to be executed.

This has the nice effect that you will not have to have comments any more.

Comment (kaxil) [reply scripts/ci/pre_commit/pre_commit_new_session_in_provide_session.py]: +1

Comment (kaxil) [reply scripts/ci/pre_commit/pre_commit_new_session_in_provide_session.py]: We should add a section in https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#coding-style-and-best-practices and print a link here

Comment (uranusjr) [reply scripts/ci/pre_commit/pre_commit_new_session_in_provide_session.py]: Actually we already do have a section on this! I’ll just link it here.

https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#database-session-handling

**Review by potiuk — APPROVED** [review]

**Comment by potiuk** [comment]
Needs rebase/conflict resolving.

**Pull request merged**